index/segment/mem/trigram_terms_dict.go

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"regexp/syntax"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+
+	"github.com/google/codesearch/index"
+)
+
+var (
+	// sentinelTrigram is stored for every field so that we can perform match all lookups.
+	sentinelTrigram = ""
```

Comment:
Would it be better to not do this? The main concern is the postings list corresponding to this entry will be pretty huge, which'd make the candidate search space for some queries large too. How do you feel about instead having the trigram dictionary return its unable to satisfy any queries needing this?

Reply:
I'm not sure. My hesitation about the latter approach is that we break the abstraction provided by the segment interface. Namely, the index would have to know what kind of queries the segment supports. For example, if I have the term `foo:c1`, the value isn't long enough to produce any trigrams and would effectively be "invisible" in the index. Thus, even if there are very few terms with key `foo`, so that the postings list for the key is small, we would still have to return an error code indicating that we are unable to satisfy the query. If we want to still support searching for such terms then we could have a mechanism in the index to account for them but then we are exposing a lot of the implementation details of the trigram index. My inclination is to go with this admittedly more heavyweight approach so that we can cleanly satisfy the segment interface and reevaluate when we have more benchmarks.